core/src/main/java/org/elasticsearch/cluster/service/ClusterService.java

```diff
@@ -244,6 +240,35 @@ protected synchronized void doStop() {
     protected synchronized void doClose() {
     }
 
+    class ClusterServiceTaskBatching extends TaskBatching {
+
+        ClusterServiceTaskBatching(Logger logger, PrioritizedEsThreadPoolExecutor threadExecutor, ThreadPool threadPool) {
+            super(logger, threadExecutor, threadPool);
+        }
+
+        @Override
+        protected void onTimeout(BatchingTask task, TimeValue timeout) {
+            ((UpdateTask) task).listener.onFailure(task.source, new ProcessClusterEventTimeoutException(timeout, task.source));
+        }
+
+        @Override
+        protected void run(Object batchingKey, List<? extends BatchingTask> tasks, String tasksSummary) {
+            ClusterStateTaskExecutor<Object> taskExecutor = (ClusterStateTaskExecutor<Object>) batchingKey;
```

Comment:
I'm doubting if we should introduce generics for the BatchingTask and batchingKey. I'm fine with keeping as is if you don't like the generic version.